Updated "Implementing JWT Authentication on Spring Boot APIs"

vladimirfomene opened the pull request:
Hello @brunokrebs, here is the updated version of the article.

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
```diff
@@ -2,13 +2,13 @@
 title: "Implementing JWT Authentication on Spring Boot APIs"
 description: "Let's learn the correct way to secure Spring Boot RESTful APIs with JWTs."
-date: 2017-08-10 18:23
+date: 2018-07-07 18:23
```
Keep the date untouched and add a new line below it with the following:

```
updated: 2018-07-17 08:30
```

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
````diff
@@ -47,14 +47,25 @@ The RESTful Spring Boot API that we are going to secure is a task list manager.
+}
 ```
 
+- Then run the unsecured RESTful API by either issuing the command `gradle bootRun` from the commandline or by
````
commandline => command line

brunokrebs replied:
here as well:

> Then, we can run the unsecured ... and running the project in _our_ favorite IDE.

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
```diff
@@ -96,31 +107,31 @@ import javax.persistence.Id;
-	public void setPassword(String password) {
-		this.password = password;
-	}
+    @Id
```
any reason why to change the code style? You probably didn't change the code style for the whole project and the whole article, have you? As such, probably, there are two code styles being used now. Please, move back to the original one.

brunokrebs replied:
It will be easier to check for the real differences.

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
```diff
@@ -133,12 +144,13 @@ This entity class contains three properties:
 public interface ApplicationUserRepository extends JpaRepository<ApplicationUser, Long> {
-	ApplicationUser findByUsername(String username);
+
+    ApplicationUser findByUsername(String username);
```
move back to the original code style (i.e. you have to use tabs instead of spaces in your code)

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
```diff
@@ -206,7 +218,7 @@ import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
 public class Application {
 
-	@Bean
+  @Bean
```
Wrong indentation. Fix your code style.

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
```diff
@@ -223,29 +235,27 @@ To support both authentication and authorization in our application, we are goin
-- create a custom implementation of [`UserDetailsService`](http://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/core/userdetails/UserDetailsService.html) to help Spring Security loading user-specific data in the framework,
-- and extend the [`WebSecurityConfigurerAdapter`](http://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/config/annotation/web/configuration/WebSecurityConfigurerAdapter.html) class to customize the security framework to our needs.
+- create a custom implementation of [`UserDetailsService`](https://docs.spring.io/spring-security/site/docs/4.2.6.RELEASE/apidocs/org/springframework/security/core/userdetails/UserDetailsService.html) to help Spring Security loading user-specific data in the framework,
+- and extend the [`WebSecurityConfigurerAdapter`](https://docs.spring.io/spring-security/site/docs/4.2.5.RELEASE/apidocs/org/springframework/security/config/annotation/web/configuration/WebSecurityConfigurerAdapter.html) class to customize the security framework to our needs.
```
I believe you are not using the `4.2.6.RELEASE` version, are you? Switch back to `current`.

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
````diff
@@ -254,60 +264,60 @@ import java.io.IOException;
 ```
 
-Note that the authentication filter that we created extends the [`UsernamePasswordAuthenticationFilter` class](http://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/web/authentication/UsernamePasswordAuthenticationFilter.html). When we add a new filter to Spring Security, we can explicitly define where in the *filter chain* we want that filter, or we can let ...
+Note that the authentication filter that we created extends the [`UsernamePasswordAuthenticationFilter` class](https://docs.spring.io/spring-security/site/docs/4.2.6.RELEASE/apidocs/org/springframework/security/web/authentication/UsernamePasswordAuthenticationFilter.html). When we add a new filter to Spring Security, we can explicitly define where in the *filter chain* we want that filter, or we  ...
````
current instead of `4.2.6.RELEASE`

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
```diff
@@ -254,60 +264,60 @@ import java.io.IOException;
 Our custom authentication filter overwrites two methods of the base class:
 
-- `attemptAuthentication`: where we parse the user's credentials and issue them to the [`AuthenticationManager`](https://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/authentication/AuthenticationManager.html).
+- `attemptAuthentication`: where we parse the user's credentials and issue them to the [`AuthenticationManager`](https://docs.spring.io/spring-security/site/docs/4.2.6.RELEASE/apidocs/org/springframework/security/authentication/AuthenticationManager.html).
```
current

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
````diff
@@ -47,14 +47,25 @@ The RESTful Spring Boot API that we are going to secure is a task list manager.
+cd springboot-auth-updated
+```
+
+- To ensure compatibility with Java 10, add the following line to the `build.gradle` of the project.
````
the whole article uses the first-person plural grammatical person. As such, this phrase should be rewritten as follows:

```
To ensure compatibility with Java 10, we have to add the following line to the `build.gradle` file:
```

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
```diff
@@ -316,163 +326,166 @@ Let's solve the missing dependency first. In the `./build.gradle` file, let's ad
+    public static final long EXPIRATION_TIME = 864_000_000; // 10 days
+    public static final String TOKEN_PREFIX = "Bearer ";
+    public static final String HEADER_STRING = "Authorization";
+    public static final String SIGN_UP_URL = "/users/sign-up";
```
code style

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
````diff
@@ -316,163 +326,166 @@ Let's solve the missing dependency first. In the `./build.gradle` file, let's ad
 ```
 
-We have extended the [`BasicAuthenticationFilter`](https://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/web/authentication/www/BasicAuthenticationFilter.html) to make Spring replace it in the *filter chain* with our custom implementation. The most important part of the filter that we've implemented is the private `getAuthentication` method. This method rea ...
+We have extended the [`BasicAuthenticationFilter`](https://docs.spring.io/spring-security/site/docs/4.2.4.RELEASE/apidocs/org/springframework/security/web/authentication/www/BasicAuthenticationFilter.html) to make Spring replace it in the *filter chain* with our custom implementation. The most important part of the filter that we've implemented is the private `getAuthentication` method. This meth ...
````
current

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
```diff
@@ -316,163 +326,166 @@ Let's solve the missing dependency first. In the `./build.gradle` file, let's ad
 
 - `configure(HttpSecurity http)`: a method where we can define which resources are public and which are secured. In our case, we set the `SIGN_UP_URL` endpoint as being public and everything else as being secured. We also configure CORS (Cross-Origin Resource Sharing) support through `http.cors()` and we add a custom security filter in the Spring Security filter chain.
-- `configure(AuthenticationManagerBuilder auth)`: a method where we defined a custom implementation of [`UserDetailsService`](https://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/core/userdetails/UserDetailsService.html) to load user-specific data in the security framework. We have also used this method to set the encrypt method used by our application (`B ...
+- `configure(AuthenticationManagerBuilder auth)`: a method where we defined a custom implementation of [`UserDetailsService`](https://docs.spring.io/spring-security/site/docs/4.2.6.RELEASE/apidocs/org/springframework/security/core/userdetails/UserDetailsService.html) to load user-specific data in the security framework. We have also used this method to set the encrypt method used by our applicati ...
```
current

vladimirfomene commented:
Hello @brunokrebs, I have made the reverted to the previous coding style. I have also only made changes on the lines the are have changed. I have also corrected the grammatical errors and date. I will be waiting for your feedback.

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
````diff
@@ -47,14 +48,24 @@ The RESTful Spring Boot API that we are going to secure is a task list manager.
+cd springboot-auth-updated
+```
+
+- To ensure compatibility with Java 10, we have to add the following line to the `build.gradle` file:
````
No need to transform this sentence into a list. That means: remove from `-` from the start.

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
````diff
@@ -47,14 +48,24 @@ The RESTful Spring Boot API that we are going to secure is a task list manager.
+}
 ```
 
+- Then, we run the unsecured RESTful API by either issuing the command `gradle bootRun` from the command line or by building and running the project in our favorite IDE.
````
No need to transform this sentence into a list. That means: remove from `-` from the start.

brunokrebs replied:
Also, I was trying to follow the article and I realized that your repo does not contain the source code inside the main directory. I guess I have mentioned it before but, anyway, move all the files that are inside the `springboot-auth-updated` directory one folder up. That is, when someone browse to https://github.com/auth0-blog/springboot-auth-updated they should see files like `build.gradle`, `gradlew`, etc.

If you don't do this, the steps described in the article won't work because you told them to clone the repo, then to `cd springboot-auth-updated`, and then to `gradle bootRun`, but there is nothing gradle can run on the root directory right now.

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
````diff
@@ -176,7 +187,7 @@ public class UserController {
 ```
 
-The implementation of the endpoint is quite simple. All it does is encrypt the password of the new user (holding it as plain text wouldn't be a good idea) and then save it to the database. The encryption process is handled by an instance of [`BCryptPasswordEncoder`](https://docs.spring.io/spring-security/site/docs/current/apidocs/org/springframework/security/crypto/bcrypt/BCryptPasswordEncoder.ht ...
+The implementation of the endpoint is quite simple. All it does is encrypt the password of the new user (holding it as plain text wouldn't be a good idea) and then save it to the database. The encryption process is handled by an instance of [`BCryptPasswordEncoder`](https://docs.spring.io/spring-security/site/docs/5.0.6.RELEASE/api/org/springframework/security/crypto/bcrypt/BCryptPasswordEncoder. ...
````
I just checked all the URLs that you have changed in this article and I understand that the previous URLs were not working anymore. However, you don't need to fix the version on the new URL, you can simply replace `5.0.6.RELEASE` with `current` on all URLs.

That is, this one would be: 

```
https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/crypto/bcrypt/BCryptPasswordEncoder.html
```

Instead of:

```
https://docs.spring.io/spring-security/site/docs/5.0.6.RELEASE/api/org/springframework/security/crypto/bcrypt/BCryptPasswordEncoder.html
```

brunokrebs commented on _posts/2017-08-10-implementing-jwt-authentication-on-spring-boot.markdown:
````diff
@@ -47,14 +48,24 @@ The RESTful Spring Boot API that we are going to secure is a task list manager.
+
 If everything works as expected, our RESTful Spring Boot API will be up and running. To test it, we can use a tool like [Postman](https://www.getpostman.com/) or [curl](https://curl.haxx.se/) to issue request to the available endpoints:
 
 ```bash
````
I am trying to follow this article but it is not working and I (probably) know why. You probably implemented the whole solution and uploaded to [the main repo of the article](https://github.com/auth0-blog/springboot-auth-updated/tree/master/springboot-auth-updated)

The problem is that you are telling readers to clone this repo and to run this repo and to issues command like `curl http://localhost:8080/tasks`. However, as you already secure the repo, these commands will get a forbidden response (HTTP 403).

What you have to do is to have a repository that contains the clean/unsecure solution so readers can follow all the steps described in the article. If you want to push the final solution to GitHub you will have to push it to another branch (for example, `final`) instead of pushing to `master`.

Please, fix that and, after fixing it, try to follow all the instructions from a new directory to check that everything is working as expected. That is, after finishing, move to a new directory, run `git clone https://github.com/auth0-blog/springboot-auth-updated.git` as described on line 51, then `cd springboot-auth-updated`, then `gradle bootRun`, etc, until you hit the end of the article.

If any command or manual task don't work it will mean you have to fix something in the article.

vladimirfomene commented:
Hello @brunokrebs, Hope you are having a great morning. I corrected the list issue in the article. I also created a new repository for the application which readers can clone and follow while reading the article. Here is the link: https://github.com/vladimirfomene/springboot-auth-updated. I have also fixed the directory structure issue that you highlighted. Finally, I also when through the article to develop the solution and it works well.

vladimirfomene commented:
Hello @brunokrebs, I have changed back the author and changed the "Application" to "SpringbootAuthUpdatedApplication" as it is in the code.
